File: app/src/main/java/com/jejefcgb/homelights/MainActivity.java
package com.jejefcgb.homelights;

import android.animation.Animator;
import android.animation.AnimatorListenerAdapter;
import android.animation.ObjectAnimator;
import android.content.DialogInterface;
import android.graphics.Color;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.View;

import com.flask.colorpicker.ColorPickerView;
import com.flask.colorpicker.builder.ColorPickerClickListener;
import com.flask.colorpicker.builder.ColorPickerDialogBuilder;
import com.github.clans.fab.FloatingActionButton;
import com.github.clans.fab.FloatingActionMenu;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import okhttp3.Call;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;

public class MainActivity extends AppCompatActivity {

    private ArrayList<Server> list;
    @BindView(R.id.my_recycler_view)
    RecyclerView mRecyclerView;

    @BindView(R.id.menu)
    FloatingActionMenu menu;

    @BindView(R.id.menu_color)
    FloatingActionButton menuColor;

    @BindView(R.id.menu_on)
    FloatingActionButton menuOn;

    @BindView(R.id.menu_off)
    FloatingActionButton menuOff;

    private RecyclerView.Adapter mAdapter;
    private RecyclerView.LayoutManager mLayoutManager;
    private final int NB_COLUMNS = 2;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        ButterKnife.bind(this);

        mRecyclerView.addItemDecoration(
                new GridSpacingItemDecoration(
                        NB_COLUMNS,
                        getResources().getDimensionPixelSize(R.dimen.default_margin),
                        true,
                        0));

        // use this setting to improve performance if you know that changes
        // in content do not change the layout size of the RecyclerView
        mRecyclerView.setHasFixedSize(true);

        // use a linear layout manager
        mLayoutManager = new GridLayoutManager(this, NB_COLUMNS);


        mRecyclerView.setLayoutManager(mLayoutManager);

        setData();


        mAdapter = new MyAdapter(list, cb);

        mRecyclerView.setAdapter(mAdapter);

    }

    private Callback cb = new Callback() {
        @Override
        public void update(List<Integer> value) {

            int menuHeight = menu.getHeight() / (menu.getChildCount() + 1);

            if (value.size() > 0 && menu.getVisibility() == View.GONE) {
                ObjectAnimator in = ObjectAnimator.ofFloat(menu, "translationY", menuHeight + 100, 0);
                in.addListener(new AnimatorListenerAdapter() {
                    @Override
                    public void onAnimationStart(Animator animation) {
                        super.onAnimationStart(animation);
                        menu.setVisibility(View.VISIBLE);
                    }
                });
                in.start();
            } else if (value.isEmpty()) {

                ObjectAnimator out = ObjectAnimator.ofFloat(menu, "translationY", 0, menuHeight + 100);
                out.addListener(new AnimatorListenerAdapter() {
                    @Override
                    public void onAnimationEnd(Animator animation) {
                        super.onAnimationEnd(animation);
                        menu.setVisibility(View.GONE);
                    }
                });
                out.start();
            }

        }
    };

    private void setData() {
        // specify an adapter (see also next example)
        list = new ArrayList<>();

        Server meubleTv = new Server();
        meubleTv.setName("Meuble TV");
        meubleTv.setIcon(R.mipmap.ic_object_tv);
        list.add(meubleTv);

        Server lit = new Server();
        lit.setName("Lit");
        lit.setIcon(R.mipmap.ic_object_bed);
        list.add(lit);

        Server bar = new Server();
        bar.setName("Bar");
        bar.setIcon(R.mipmap.ic_object_bar);
        list.add(bar);

        Server desk = new Server();
        desk.setName("Bureau");
        desk.setIcon(R.mipmap.ic_object_desk);
        list.add(desk);

    }

    @OnClick(R.id.menu_color)
    void openColorPicker() {

        if (menu.isOpened()) {
            menu.close(true);
        }


        ColorPickerDialogBuilder
                .with(MainActivity.this)
                .setTitle(getString(R.string.color_title))
                .wheelType(ColorPickerView.WHEEL_TYPE.FLOWER)
                .density(12)
                .showAlphaSlider(false)
                .setPositiveButton(getString(R.string.validate), new ColorPickerClickListener() {
                    @Override
                    public void onClick(DialogInterface dialogInterface, int i, Integer[] integers) {


                        ((MyAdapter) mAdapter).resetSelectedPos();
                        OkHttpClient client = new OkHttpClient();


                        String url  = String.format("http://192.168.1.41:8080/api/lights/%1$d/%2$d/%3$d",Color.red(i), Color.green(i), Color.blue(i));

                        Request request = new Request.Builder()
                                .url(url)
                                .build();

                        client.newCall(request).enqueue(new okhttp3.Callback() {
                            @Override
                            public void onFailure(Call call, IOException e) {

                            }

                            @Override
                            public void onResponse(Call call, Response response) throws IOException {

                            }
                        });
                    }
                })
                        // TODO Send commands
                .setNegativeButton(getString(R.string.cancel), (dialog, which) -> dialog.dismiss())
                .build()
                .show();
    }

    @OnClick(R.id.menu_off)
    void shutDown(){
        ((MyAdapter)mAdapter).resetSelectedPos();
        OkHttpClient client = new OkHttpClient();


        String url  = String.format("http://192.168.1.41:8080/api/lights/0/0/0");

        Request request = new Request.Builder()
                .url(url)
                .build();

        client.newCall(request).enqueue(new okhttp3.Callback() {
            @Override
            public void onFailure(Call call, IOException e) {

            }

            @Override
            public void onResponse(Call call, Response response) throws IOException {

            }
        });
    }

}

File: app/src/main/java/com/jejefcgb/homelights/MainActivity.kt
package com.jejefcgb.homelights

import android.graphics.Color
import android.os.Bundle
import android.view.View
import android.widget.Spinner
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import butterknife.BindView
import butterknife.ButterKnife
import butterknife.OnClick
import com.afollestad.materialdialogs.MaterialDialog
import com.afollestad.materialdialogs.customview.customView
import com.afollestad.materialdialogs.customview.getCustomView
import com.jejefcgb.homelights.HomeLightsApplication.Companion.TYPE_ROOM
import com.jejefcgb.homelights.HomeLightsApplication.Companion.config
import com.jejefcgb.homelights.data.model.Furniture
import com.jejefcgb.homelights.data.model.Home
import com.jejefcgb.homelights.data.model.Room
import com.jejefcgb.homelights.ui.GridSpacingItemDecoration
import com.jejefcgb.homelights.ui.HomeAdapter
import com.jejefcgb.homelights.utils.APIHelper
import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.android.synthetic.main.my_custom_view.view.*
import okhttp3.OkHttpClient
import okhttp3.Request


class MainActivity : AppCompatActivity() {

    @BindView(R.id.main_recycler_view)
    lateinit var mRecyclerView: RecyclerView

    private lateinit var mAdapter: HomeAdapter
    private lateinit var mLayoutManager: RecyclerView.LayoutManager


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        ButterKnife.bind(this)

        // Set up recycler view
        mRecyclerView.addItemDecoration(
                GridSpacingItemDecoration(
                        NB_COLUMNS,
                        resources.getDimensionPixelSize(R.dimen.default_margin),
                        true,
                        0))

        mRecyclerView.setHasFixedSize(true)
        mLayoutManager = GridLayoutManager(this, NB_COLUMNS)

        mRecyclerView.layoutManager = mLayoutManager
        mAdapter = HomeAdapter(this, config.rooms, TYPE_ROOM)
        mRecyclerView.adapter = mAdapter

        // Swipe to refresh
        swipe_container.setOnRefreshListener { getRemoteConfig() }
        swipe_container.setColorSchemeColors(Color.BLUE, Color.RED, Color.GREEN)

        // Action bar
        supportActionBar?.setDisplayShowHomeEnabled(true)

        getRemoteConfig()

    }


    private fun getRemoteConfig() {

        swipe_container.isRefreshing = true
        val client = OkHttpClient()
        val request = Request.Builder()
                .url("${APIHelper.API_ADDRESS}/config")
                .build()

        //FIXME :  TO REMOVE
        swipe_container.isRefreshing = false
        config = Home("home", arrayListOf(
                Room(0, "Salon", "ic_object_tv", listOf(
                        Furniture(0, "Verres", "192.168.1.200", "ic_object_glasses")
                )),
                Room(1, "Chambre du naze", "ic_object_bed", listOf(
                        Furniture(1, "Lit", "192.168.200.1", "ic_object_bed"),
                        Furniture(2, "Bureau", "192.168.200.1", "ic_object_desk")
                )),
                Room(2, "Chambre de la naze", "ic_object_bed", listOf(
                        Furniture(3, "Etagère", "192.168.1.201", "ic_object_shelf"),
                        Furniture(4, "Support écran", "192.168.200.1", "ic_object_tv")
                ))))

        mAdapter.data = config.rooms
        mAdapter.notifyDataSetChanged()

        Toast.makeText(this@MainActivity, "FIXME", Toast.LENGTH_SHORT).show()
        // FIXME : END TO REMOVE

//        client.newCall(request).enqueue(object : okhttp3.Callback {
//            override fun onFailure(call: Call, e: IOException) {
//                this@MainActivity.runOnUiThread {
//                    swipe_container.isRefreshing = false
//
//                    toast("Impossible de récupérer la configuration. Veuillez vérifier votre réseau wifi")
//
//                    config = Home("home", listOf(
//                            Room(0, "Salon", "ic_object_tv", listOf(
//                                    Furniture("Bar", "192.168.200.1", "ic_object_bar")
//                            )),
//                            Room(1, "Chambre du naze", "ic_object_bed", listOf(
//                                    Furniture("Lit", "192.168.200.1", "ic_object_bed"),
//                                    Furniture("Bureau", "192.168.200.1", "ic_object_desk")
//                            )),
//                            Room(2, "Chambre de la naze", "ic_object_bed", listOf(
//                                    Furniture("Etagère", "192.168.200.1", "ic_object_shelf"),
//                                    Furniture("Support écran", "192.168.200.1", "ic_object_tv")
//                            ))))
//
//                    mAdapter.notifyDataSetChanged()
//                }
//            }
//
//            @Throws(IOException::class)
//            override fun onResponse(call: Call, response: Response) {
//                this@MainActivity.runOnUiThread {
//                    swipe_container.isRefreshing = false
//
//                    if (response.isSuccessful) {
//
//                        val jsonAdapter = Moshi.Builder().build().adapter(Home::class.java)
//                        val json = response.body()?.string()
//
//                        HomeLightsApplication.config = jsonAdapter.fromJson(json as String) as Home //TODO : vérifier comportement si un champ manquant
//                        mAdapter.notifyDataSetChanged()
//
//                        toast("Configuration chargée")
//                    } else {
//                        val body = response.body()?.string()
//                        toast("Une erreur est survenue lors de la récupération de la configuration : (${response.code()}) > $body")
//                    }
//                }
//            }
//        })
    }

    @OnClick(R.id.main_fab)
    fun addRoom() {
        val dialog = MaterialDialog(this).show {
            customView(R.layout.my_custom_view)
            positiveButton(text = "Ajouter")
            negativeButton(text = "Annuler")
        }

        val v: View = dialog.getCustomView()
        val spinner: Spinner = v.custom_spinner
        spinner.adapter = IconAdapter(this)

        val nameEt : TextView = v.custom_et_name

        dialog.negativeButton { it.dismiss() }
        dialog.positiveButton {
            val icon = resources.getResourceEntryName(spinner.adapter.getItem(spinner.selectedItemPosition) as Int)
            val id: Int = config.rooms.map { x -> x.id }.max() as Int + 1
            val name = nameEt.text.toString().trim()

            config.rooms.add(Room(
                    id,
                    name,
                    icon,
                    ArrayList()
            ))

            mAdapter.notifyDataSetChanged()

        }


    }

    private fun toast(message: String) {
        Toast.makeText(this@MainActivity, message, Toast.LENGTH_SHORT).show()
    }

    companion object {
        internal const val NB_COLUMNS = 1
    }

}
